### PR TITLE
Configurable Geocoder (fix #210)

### LINKED ISSUE #210: Configurable geocoder
At the moment GHC is using freegeoip.net service to translate ip to location. This is 3rd party service, and can return errors unexpectedly. GHC could have pluggable geocoder backends, so other options (local copy of GeoIP from MaxMind for example) could be used too.

### PR DESCRIPTION
**Pull request by borrob**
An implementation for issue #210: Configurable Geocoder

The plugin system is used to create a configurable geocoder. The geocoder plugin class and specific parameters are specified in the `config_main.py` or `config_site.py`. The base setup refers to the ip-api.com that is already used at the moment.

This PR provides three plugins: using a geocode service on the web via a GET request and via a POST request. The third geocoder returns a (configurable) fixed location which might be useful for intranet/on prem installations of GHC.

I did not check compatibility with the other two open PRs, but conflicts (if any) should be minor.

### CODE REVIEW

**Comment by justb4** [comment]
Need some time to dive in. At least the build failure is not from this PR, but a test that Probes the Resource https://demo.pygeoapi.io/master. That Resource has been failing I now see since jan 28, 2020 10:55 GMT. See https://demo.geohealthcheck.org/resource/165
Latest commit of GHC `master` and thus `demo.geohealthcheck.org` instance is from dec 2019. Need to investigate. The error signaled is `Validate OpenAPI Compliance:HTTP Error 404: Not Found`. 

Dived into this, was quite a route among various products/components. All described here: https://github.com/geopython/pygeoapi/issues/356...also to note @tomkralidis

**Comment by justb4** [comment]
https://github.com/opengeospatial/wps-rest-binding/issues/63 was cause and fixed so build shoild pass now.

**Comment by justb4** [comment]
Travis is IMO failing as it sets up for Ubuntu Xenial (16.4) which does not provide GDAL v3 via Ubuntugis unstable. .Travis file needs to upgrade to Bionic (18.4) as to enable GDAL v3, see also https://docs.travis-ci.com/user/reference/bionic/ and Ubuntugis PPA info: https://launchpad.net/~ubuntugis/+archive/ubuntu/ubuntugis-unstable .

**Comment by borrob** [comment]
Great find and I am happy it all works now. All tests pass.

File: docs/install.rst [hunk 1]
```diff
@@ -81,6 +81,7 @@ Install
   # - GHC_REQUIRE_WEBAPP_AUTH  # optional: to require authentication to access webapp
   # - GHC_SMTP  # if GHC_NOTIFICATIONS is enabled
   # - GHC_MAP  # or use default settings
+  # - GEOIP  # or use the feault settings
```
Comment (justb4) [inline docs/install.rst:84]: typo: "default"

File: docs/config.rst [hunk 2]
```diff
@@ -42,6 +42,7 @@ The configuration options are:
   - **centre_long**: Centre longitude for homepage map
   - **maxzoom**: maximum zoom level
   - **subdomains**: available subdomains to help with parallel requests
+- **GEOIP**: configuration for the geolocater service plugin. Default is the ip-api.com api.
```
Comment (justb4) [inline docs/config.rst:45]: `geolocater` should be `geolocator`, multiple files.

File: docs/plugins.rst [hunk 3]
```diff
@@ -10,7 +10,8 @@ one may want to perform. Hence developers can extend or even replace
 the GHC standard Plugins with custom implementations.
 
 Two Plugin types exist that can be extended: the `Probe` and `Check` class.
-In v0.7.0 also plugins for Resource Authentication, `ResourceAuth`, were added.
+In v0.7.0 also plugins for Resource Authentication, `ResourceAuth`, were added
+and in v0.8.0 the geocoder plugin was introduced.
```
Comment (justb4) [inline docs/plugins.rst:14]: will hopefully be 0.8.1

**Review by justb4 — CHANGES_REQUESTED** [review]
Good to see the Plugin system used! 
Main comment/changes needed: any config change also requires changes to the config for Docker deployment. Not handy but Docker uses env-vars a lot. See the files:
[config_site.py](https://github.com/geopython/GeoHealthCheck/blob/master/docker/config_site.py) and default env vars in [Dockerfile](https://github.com/geopython/GeoHealthCheck/blob/master/Dockerfile). Changes will be bit similar to SMTP config convention.

File: docs/config.rst (hunk 2 above)
Comment (borrob) [reply docs/config.rst:45]: I feel stupid.

**Comment by borrob** [comment]
Thanks for the review work!  I'll try to fix things up, but I'm quite busy lately, so it might take some time.

**Comment by justb4** [comment]
@borrob no sweat! merge with current master, and take your time!

File: docs/install.rst (hunk 1 above)
Comment (borrob) [reply docs/install.rst]: Fixed in 9699493

File: docs/config.rst (hunk 2 above)
Comment (borrob) [reply docs/config.rst]: Fixed with: e8868a1

File: docs/plugins.rst (hunk 3 above)
Comment (borrob) [reply docs/plugins.rst]: Made a reference to 0.9 with commit 4fa7da7

**Comment by borrob** [comment]
> Good to see the Plugin system used!
> Main comment/changes needed: any config change also requires changes to the config for Docker deployment. Not handy but Docker uses env-vars a lot. See the files:
> [config_site.py](https://github.com/geopython/GeoHealthCheck/blob/master/docker/config_site.py) and default env vars in [Dockerfile](https://github.com/geopython/GeoHealthCheck/blob/master/Dockerfile). Changes will be bit similar to SMTP config convention.

I fixed the easy ones. I'll start to check up on the config changes.

**Comment by borrob** [comment]
> > Good to see the Plugin system used!
> > Main comment/changes needed: any config change also requires changes to the config for Docker deployment. Not handy but Docker uses env-vars a lot. See the files:
> > [config_site.py](https://github.com/geopython/GeoHealthCheck/blob/master/docker/config_site.py) and default env vars in [Dockerfile](https://github.com/geopython/GeoHealthCheck/blob/master/Dockerfile). Changes will be bit similar to SMTP config convention.
> 
> I fixed the easy ones. I'll start to check up on the config changes.

@justb4 Can you be a bit more specific? I changed the `config_main.py` [here](https://github.com/borrob/GeoHealthCheck/blob/4fa7da7f1b0f13d225ea96c8e5de661c584a5d1a/GeoHealthCheck/config_main.py#L180), the config for the docker file `config_site.py` [here](https://github.com/borrob/GeoHealthCheck/blob/4fa7da7f1b0f13d225ea96c8e5de661c584a5d1a/docker/config_site.py#L108), and it is described in the documentation. Or would you rather have it configured in the Docker env-var ([Dockerfile](https://github.com/borrob/GeoHealthCheck/blob/geocoder/Dockerfile) and overruled in [ghc.env](https://github.com/borrob/GeoHealthCheck/blob/geocoder/docker/compose/ghc.env) for example)?

**Comment by justb4** [comment]
@borrob the reason for the specific Docker var setup is to enable overruling the default GHC configuration in `config_site.py` via env vars. For example when deploying in Kubernetes or other environments where there is no docker-compose nor `ghc.env` local Docker volume-mount possible. 

The convention would be the same as for SMTP. Given the current GEOIP config this would require 3 env vars, e.g.
`GHC_GEOIP_URL`, `GHC_GEOIP_LATFIELD`,  `GHC_GEOIP_LONFIELD`.

Set defaults in the Dockerfile:
```
ENV LC_ALL="en_US.UTF-8" \
.
.
	GHC_GEOIP_URL='http://ip-api.com/json/{hostname}' \
	GHC_GEOIP_LATFIELD='lat' \
	GHC_GEOIP_LONFIELD='lon' \
.
.
```

And then in the `config_site.py` use values from environment (either the default from the Dockerfile or an overruled value):

```
GEOIP = {
    'plugin': 'GeoHealthCheck.plugins.geocode.webgeocoder.HttpGetGeocoder',
    'parameters': {
        'geocoder_url':  os.environ['GHC_GEOIP_URL'],
        'lat_field':  os.environ['GHC_GEOIP_LATFIELD'],
        'lon_field':  os.environ['GHC_GEOIP_LONFIELD']
}
```

Feels like duplicate work, but this has evolved historically and came from Kubernetes deployment. This convention is very common for many tools. Hence also the convention to start env varnames with `GHC_`

**Comment by justb4** [comment]
Looks good now! I propose to first finalize #349 and I will merge #347 and then this one.

File: Dockerfile [hunk 4]
```diff
@@ -44,6 +44,9 @@ ENV LC_ALL="en_US.UTF-8" \
 	GHC_SMTP_SSL=False \
 	GHC_SMTP_USERNAME=None \
 	GHC_SMTP_PASSWORD=None \
+    GHC_GEOIP_URL='http://ip-api.com/json/{hostname}' \
```
Comment (tomkralidis) [inline Dockerfile:47]: Can we align/indent accordingly?

File: GeoHealthCheck/geocoder.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+from plugin import Plugin
```
Comment (tomkralidis) [inline GeoHealthCheck/geocoder.py:1]: Can we add a license header?

File: GeoHealthCheck/plugins/geocode/fixedlocation.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+from GeoHealthCheck.geocoder import Geocoder
```
Comment (tomkralidis) [inline GeoHealthCheck/plugins/geocode/fixedlocation.py:1]: License header here as well

File: GeoHealthCheck/plugins/geocode/webgeocoder.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
+import requests
```
Comment (tomkralidis) [inline GeoHealthCheck/plugins/geocode/webgeocoder.py:1]: License header

**Review by tomkralidis — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 4 above)
Comment (borrob) [reply Dockerfile:47]: Aaarrgh. tab vs space discussion.
Fixed with 39183ac

File: GeoHealthCheck/geocoder.py (hunk 5 above)
Comment (borrob) [reply GeoHealthCheck/geocoder.py:1]: Fixed with 7bdf022

File: GeoHealthCheck/plugins/geocode/fixedlocation.py (hunk 6 above)
Comment (borrob) [reply GeoHealthCheck/plugins/geocode/fixedlocation.py:1]: Fixed with 7bdf022

File: GeoHealthCheck/plugins/geocode/webgeocoder.py (hunk 7 above)
Comment (borrob) [reply GeoHealthCheck/plugins/geocode/webgeocoder.py:1]: Fixed with 7bdf022

**Comment by borrob** [comment]
Thanks for all the review work and the explanation on the docker variables.

File: GeoHealthCheck/geocoder.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        self._geocoder_vars = geocoder_vars
+
+    def locate(self, hostname):
+        """
+        Perform a locate on the host.
+
+        :param hostname string: the hostname of the server for which we want
+                                the coords.
+
+        TOOD: return result as tuple with locatin in lat-lon like: (52.4, 21.0)
```
Comment (justb4) [inline GeoHealthCheck/geocoder.py:55]: typos:
TOOD --> TODO
locatin --> location

**Comment by justb4** [comment]
Ok,  PRs #349  #347 are now in. Maybe you want to merge those in to get the green light on tests first.
So AFAICT a default config is in place, so upgrades should be smooth. You have also tested the Docker image?

BTW Whenever we merge in master branch the Docker image `latest` is automatically built and deployed as  https://demo.geohealthcheck.org .

Need some time for full review...

File: GeoHealthCheck/geocoder.py (hunk 8 above)
Comment (borrob) [reply GeoHealthCheck/geocoder.py]: I should pay more attention to my typing.
Fixed with 0c377b0

**Comment by borrob** [comment]
> Ok, PRs #349 #347 are now in. Maybe you want to merge those in to get the green light on tests first.
> So AFAICT a default config is in place, so upgrades should be smooth. You have also tested the Docker image?
> 
> BTW Whenever we merge in master branch the Docker image `latest` is automatically built and deployed as https://demo.geohealthcheck.org .
> 
> Need some time for full review...

Great!
I merged to new master to this PR (e200a8a) and I should get `flake8` into my workflow before I commit.
I ran all the tests and I did make a (local) docker build. The app runs and all seems fine...

**Comment by justb4** [comment]
Thanks @borrob !

**Pull request merged**